test/wasm_test.cc

```diff
@@ -0,0 +1,99 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/proxy-wasm/wasm.h"
+
+#include "gtest/gtest.h"
+
+#include "test/utility.h"
+
+namespace proxy_wasm {
+
+auto test_values = testing::ValuesIn(getRuntimes());
+
+INSTANTIATE_TEST_SUITE_P(Runtimes, TestVM, test_values);
+
+// Failcallbacks only used for runtimes - not available for nullvm.
+TEST_P(TestVM, GetOrCreateThreadLocalWasmFailCallbacks) {
+  const auto plugin_name = "plugin_name";
+  const auto root_id = "root_id";
+  const auto vm_id = "vm_id";
+  const auto plugin_config = "plugin_config";
+  const auto fail_open = false;
+
+  // Create a plugin.
+  const auto plugin = std::make_shared<PluginBase>(plugin_name, root_id, vm_id, runtime_,
+                                                   plugin_config, fail_open, "plugin_key");
+
+  // Define callbacks.
+  WasmHandleFactory wasm_handle_factory =
+      [this](std::string_view vm_key) -> std::shared_ptr<WasmHandleBase> {
+    auto base_wasm = std::make_shared<WasmBase>(newVm(), "vm_id", "vm_config", vm_key,
+                                                std::unordered_map<std::string, std::string>{},
+                                                AllowedCapabilitiesMap{});
+    return std::make_shared<WasmHandleBase>(base_wasm);
+  };
+
+  WasmHandleCloneFactory wasm_handle_clone_factory =
+      [this](std::shared_ptr<WasmHandleBase> base_wasm_handle) -> std::shared_ptr<WasmHandleBase> {
+    auto wasm = std::make_shared<WasmBase>(base_wasm_handle,
+                                           [this]() -> std::unique_ptr<WasmVm> { return newVm(); });
+    return std::make_shared<WasmHandleBase>(wasm);
+  };
+
+  PluginHandleFactory plugin_handle_factory =
+      [](std::shared_ptr<WasmHandleBase> base_wasm,
+         std::shared_ptr<PluginBase> plugin) -> std::shared_ptr<PluginHandleBase> {
+    return std::make_shared<PluginHandleBase>(base_wasm, plugin);
+  };
+
+  // Read the minimal loadable binary.
+  auto source = readTestWasmFile("abi_export.wasm");
+
+  // Create base Wasm via createWasm.
+  auto base_wasm_handle =
+      createWasm("vm_key", source, plugin, wasm_handle_factory, wasm_handle_clone_factory, false);
+  ASSERT_TRUE(base_wasm_handle && base_wasm_handle->wasm());
+
+  // Create a thread local plugin.
+  auto thread_local_plugin = getOrCreateThreadLocalPlugin(
+      base_wasm_handle, plugin, wasm_handle_clone_factory, plugin_handle_factory);
+  ASSERT_TRUE(thread_local_plugin && thread_local_plugin->plugin());
+
+  // Cause runtime crash.
+  thread_local_plugin->wasm()->wasm_vm()->fail(FailState::RuntimeError, "runtime error msg");
+  ASSERT_TRUE(thread_local_plugin->wasm()->isFailed());
+
+  // Create another thread local plugin with the same configuration.
+  // This one should not end up using the failed VM.
```

Comment:
Could you add a check that verifies that it's indeed a different VM?

Reply:
https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/181/commits/a350a499cc3d668b7f3b91b8ee99421ce15889b5 done.

Reply:
also added more assertion to check if the non-failed VM is actually reused: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/181/commits/b0f0df372a04974de7dde052fbad8368eb439493